### PR TITLE
fix: make devfed immediately kill processes when exiting

### PR DESCRIPTION
**Pull request by douglaz**
Improves a lot https://github.com/fedimint/fedimint/issues/2838

Will make devimint always shut down the subprocesses when devfed exits (e.g when tmuxinator/mprocs is closed).

### CODE REVIEW

**Review by elsirion — APPROVED** [review]

File: devimint/src/main.rs [hunk 1]
```diff
@@ -1114,59 +1114,93 @@ async fn setup(arg: CommonArgs) -> Result<(ProcessManager, TaskGroup)> {
...
 }
 
 #[tokio::main]
 async fn main() -> Result<()> {
     let args = Args::parse();
     match args.command {
         Cmd::ExternalDaemons => {
-            let (process_mgr, task_group) = setup(args.common).await?;
+            let process_mgr = setup(args.common).await?;
+            let task_group = TaskGroup::new();
```
Comment (okjodom) [inline devimint/src/main.rs:1126]: @douglaz, I'm trying to understand the implication of this refactor for creating TG and installing kill handler. Does moving these two lines from within setup have a specific effect in killing processes later?

Comment (okjodom) [reply devimint/src/main.rs:1126]: oh maybe it's because tg is only used in `DevFed` and `ExternalDaemons` scenario?

**Review by okjodom — APPROVED** [review]
I had a curiosity question, otherwise, lgtm.
Landing to get the fresh benefits in `fedimint/ui`

File: devimint/src/main.rs [hunk 2]
```diff
@@ -1114,59 +1114,93 @@ async fn setup(arg: CommonArgs) -> Result<(ProcessManager, TaskGroup)> {
...
         }
         Cmd::DevFed => {
-            let (process_mgr, task_group) = setup(args.common).await?;
-            let dev_fed = dev_fed(&process_mgr).await?;
-            dev_fed.fed.pegin(10_000).await?;
-            dev_fed.fed.pegin_gateway(20_000, &dev_fed.gw_cln).await?;
-            dev_fed.fed.pegin_gateway(20_000, &dev_fed.gw_lnd).await?;
-            let _daemons = write_ready_file(&process_mgr.globals, Ok(dev_fed)).await?;
-            task_group.make_handle().make_shutdown_rx().await.await?;
+            let process_mgr = setup(args.common).await?;
```
Comment (justinmoon) [inline devimint/src/main.rs:1134]: The PR generally looks good, but this code block is a little hard to follow.

File: devimint/src/main.rs [hunk 3]
```diff
@@ -1114,59 +1114,93 @@ async fn setup(arg: CommonArgs) -> Result<(ProcessManager, TaskGroup)> {
...
+            let task_group = TaskGroup::new();
```
Comment (dpc) [inline devimint/src/main.rs:1135]: Maybe we should have `TaskGroup::new_with_kill_handler()` to saves some typing. :D

File: devimint/src/util.rs [hunk 4]
```diff
@@ -21,19 +22,16 @@ fn kill(child: &Child) {
...
-        let mut child = std::mem::take(&mut process_handle_inner.child).unwrap();
-        info!(LOG_DEVIMINT, "killing {}", process_handle_inner.name);
-        kill(&child);
-        child.wait().await?;
+    pub async fn kill(&self) -> Result<()> {
+        let mut inner = self.0.lock().await;
+        if let Some(mut child) = inner.child.take() {
+            info!(LOG_DEVIMINT, "killing {}", inner.name);
+            kill(&child);
+            child.wait().await?;
```
Comment (dpc) [inline devimint/src/util.rs:33]: We should ignore the error here. If we're killing it, chances are high it will fail, and we don't really care anymore.

File: devimint/src/main.rs (hunk 1 above)
Comment (douglaz) [reply devimint/src/main.rs:1126]: Yes, it's not used everywhere and I think being explicit is better if you are reading the code for the first time

File: devimint/src/util.rs (hunk 4 above)
Comment (dpc) [reply devimint/src/util.rs:33]: We probably shouldn't even wait here in the first place. We've sent it a kill signal, we're done.

File: devimint/src/util.rs [hunk 5]
```diff
@@ -79,10 +82,19 @@ impl ProcessManager {
...
             child: Some(child),
-        })))
+        })));
+        self.handles.lock().await.push(handle.clone());
+        Ok(handle)
+    }
+
+    pub async fn kill_all_children(&self) -> Result<()> {
+        for handle in self.handles.lock().await.iter() {
+            handle.kill().await?;
```
Comment (dpc) [inline devimint/src/util.rs:95]: This terminating early could prevent us from killing other children. Biggest issue I see.

File: devimint/src/util.rs (hunk 4 above)
Comment (douglaz) [reply devimint/src/util.rs:33]: No, we must be sure they are dead :wink:

File: devimint/src/main.rs (hunk 2 above)
Comment (douglaz) [reply devimint/src/main.rs:1134]: To understand what is happening read the log messages

Comment (douglaz) [reply devimint/src/main.rs:1165]: Also added to `run-ui` command.

File: devimint/src/main.rs (hunk 3 above)
Comment (douglaz) [reply devimint/src/main.rs:1166]: For me it's fine the way it is.

File: devimint/src/util.rs (hunk 5 above)
Comment (douglaz) [reply devimint/src/util.rs]: Now wil kill everything in parallel and just log the resulting errors

**Comment by douglaz** [comment]
Addressed most of comments

File: devimint/src/util.rs (hunk 4 above)
Comment (dpc) [reply devimint/src/util.rs:33]: But the wait here doesn't not check if killed them right. It checks the return status of the command itself, which will be an error ... because we killed it. Unless it was already dead. No?

Comment (douglaz) [reply devimint/src/util.rs:33]: It may or may be not an error. We are sending a SIGTERM, so the process can gracefully shutdown.

Comment (douglaz) [reply devimint/src/util.rs:33]: If the process just ignores SIGTERM, then this won't return

Comment (douglaz) [reply devimint/src/util.rs:33]: we could implement a timeout then retry will SIGKILL, but our processes are well behaved, so I don't think we need this for now

Comment (dpc) [reply devimint/src/util.rs:33]: Oh. You're right. I confused myself and thought we're sending SIGKILL already.

But I still think that most programs will terminate with an non-zero code on any signal:

image

(red `>` means non-zero exit value, Ctrl+C sends a SIGINT).

So we should wait, but is there a point in making in an `Err`?

File: devimint/src/util.rs [hunk 6]
```diff
@@ -52,13 +50,18 @@ impl Drop for ProcessHandleInner {
     }
 }
 
+#[derive(Clone)]
 pub struct ProcessManager {
-    pub globals: vars::Global,
+    pub globals: Arc<vars::Global>,
+    handles: Arc<Mutex<Vec<ProcessHandle>>>,
```
Comment (Maan2003) [inline devimint/src/util.rs:56]: incorrect: keeping a reference to ProcessHandle will prevent it from stopping on drop.

Comment (douglaz) [reply devimint/src/util.rs:56]: So what? Shouldn't we always explicit kill them?

Comment (dpc) [reply devimint/src/util.rs:56]: Should we make it a `Vec<WeakProcessHandle>` (with `Weak` instead of `Arc`?)

Comment (Maan2003) [reply devimint/src/util.rs:56]: I liked the invariant that Bitcoind value existing means bitcoind is running, maybe it is not important?

**Comment by Maan2003** [comment]
Sorry misclick on disabling auto merge

**Comment by douglaz** [comment]
WASM tests are failing :cry:

File: devimint/src/main.rs [hunk 7]
```diff
@@ -1114,58 +1114,106 @@ async fn setup(arg: CommonArgs) -> Result<(ProcessManager, TaskGroup)> {
...
+async fn cleanup_on_exit(
+    main_process: impl futures::Future<Output = Result<()>>,
+    task_group: TaskGroup,
+    process_mgr: ProcessManager,
+) -> anyhow::Result<()> {
+    // This select makes it possible to exit earlier if a signal is received before
+    // the main process is finished
+    tokio::select! {
+        _ = task_group.make_handle().make_shutdown_rx().await => {
+            info!("Received shutdown signal before finishing setup");
```
Comment (Maan2003) [inline devimint/src/main.rs:1129]: not sure about "Setup". Should be "main process"?
same for rest of logs in this function.

**Review by Maan2003 — COMMENTED** [review]
LGTM, minor nits

**Comment by douglaz** [comment]
Closed in favor of https://github.com/fedimint/fedimint/pull/2873
